File: aws/ses/typescript/src/http/response.ts
/**
 * HTTP response parsing for AWS SES API.
 *
 * This module provides utilities for parsing AWS SES API responses,
 * extracting pagination tokens, and handling errors.
 *
 * @module http/response
 */

import type { HttpResponse, PaginatedResponse, AwsErrorResponse } from './types.js';

/**
 * Parse a JSON response body.
 *
 * This is a type-safe wrapper around JSON.parse that provides
 * better error messages for invalid JSON.
 *
 * @template T - The expected response type
 * @param body - Response body as a string
 * @returns Parsed JSON object
 * @throws Error if JSON parsing fails
 *
 * @example
 * ```typescript
 * interface SendEmailResponse {
 *   MessageId: string;
 * }
 *
 * const body = '{"MessageId":"msg-123"}';
 * const response = parseResponse<SendEmailResponse>(body);
 * console.log('Message ID:', response.MessageId);
 * ```
 */
export function parseResponse<T>(body: string): T {
  if (!body || body.trim() === '') {
    throw new Error('Response body is empty');
  }

  try {
    return JSON.parse(body) as T;
  } catch (error) {
    throw new Error(
      `Failed to parse JSON response: ${error instanceof Error ? error.message : String(error)}`
    );
  }
}

/**
 * Extract the NextToken from a paginated response.
 *
 * Many AWS SES API operations return paginated results with a NextToken
 * field. This function safely extracts that token.
 *
 * @param response - Parsed response object
 * @returns Next token if present, undefined otherwise
 *
 * @example
 * ```typescript
 * const response = {
 *   Identities: [
 *     { IdentityName: 'example.com', IdentityType: 'DOMAIN' }
 *   ],
 *   NextToken: 'abc123'
 * };
 *
 * const nextToken = extractNextToken(response);
 * console.log('Next token:', nextToken); // 'abc123'
 * ```
 */
export function extractNextToken(response: unknown): string | undefined {
  if (typeof response !== 'object' || response === null) {
    return undefined;
  }

  const obj = response as Record<string, unknown>;

  // Check for NextToken field (AWS SES v2 API convention)
  if (typeof obj['NextToken'] === 'string') {
    return obj['NextToken'];
  }

  // Check for nextToken field (alternative convention)
  if (typeof obj['nextToken'] === 'string') {
    return obj['nextToken'];
  }

  return undefined;
}

/**
 * Parse a paginated response.
 *
 * This function parses a response that contains both items and a pagination token.
 *
 * @template T - The type of items in the response
 * @param body - Response body as a string
 * @param itemsKey - The key containing the items array (default: 'Items')
 * @returns Paginated response with items and optional next token
 * @throws Error if parsing fails or items key is not found
 *
 * @example
 * ```typescript
 * interface Identity {
 *   IdentityName: string;
 *   IdentityType: string;
 * }
 *
 * const body = JSON.stringify({
 *   Identities: [
 *     { IdentityName: 'example.com', IdentityType: 'DOMAIN' }
 *   ],
 *   NextToken: 'abc123'
 * });
 *
 * const result = parsePaginatedResponse<Identity>(body, 'Identities');
 * console.log('Items:', result.items);
 * console.log('Next token:', result.nextToken);
 * ```
 */
export function parsePaginatedResponse<T>(
  body: string,
  itemsKey = 'Items'
): PaginatedResponse<T> {
  const parsed = parseResponse<Record<string, unknown>>(body);

  // Extract items array
  const items = parsed[itemsKey];
  if (!Array.isArray(items)) {
    throw new Error(`Response does not contain a valid '${itemsKey}' array`);
  }

  // Extract next token
  const nextToken = extractNextToken(parsed);

  return {
    items: items as T[],
    nextToken,
  };
}

/**
 * Check if an HTTP response indicates success.
 *
 * @param response - HTTP response
 * @returns true if status is in 2xx range
 *
 * @example
 * ```typescript
 * const response: HttpResponse = {
 *   status: 200,
 *   headers: {},
 *   body: '{"success":true}'
 * };
 *
 * if (isSuccessResponse(response)) {
 *   console.log('Request succeeded');
 * }
 * ```
 */
export function isSuccessResponse(response: HttpResponse): boolean {
  return response.status >= 200 && response.status < 300;
}

/**
 * Check if an HTTP response indicates an error.
 *
 * @param response - HTTP response
 * @returns true if status is 4xx or 5xx
 *
 * @example
 * ```typescript
 * const response: HttpResponse = {
 *   status: 400,
 *   headers: {},
 *   body: '{"__type":"ValidationException"}'
 * };
 *
 * if (isErrorResponse(response)) {
 *   console.error('Request failed');
 * }
 * ```
 */
export function isErrorResponse(response: HttpResponse): boolean {
  return response.status >= 400;
}

/**
 * Parse an AWS error response.
 *
 * AWS returns errors in a specific JSON format. This function extracts
 * the error type and message from the response.
 *
 * @param response - HTTP response
 * @returns Parsed error information
 *
 * @example
 * ```typescript
 * const response: HttpResponse = {
 *   status: 400,
 *   headers: {},
 *   body: JSON.stringify({
 *     __type: 'MessageRejected',
 *     message: 'Email address is not verified'
 *   })
 * };
 *
 * const error = parseAwsError(response);
 * console.log('Error type:', error.type);
 * console.log('Error message:', error.message);
 * console.log('Is retryable:', error.retryable);
 * ```
 */
export function parseAwsError(response: HttpResponse): AwsErrorInfo {
  let errorResponse: AwsErrorResponse | undefined;

  // Try to parse error response as JSON
  if (response.body) {
    try {
      errorResponse = parseResponse<AwsErrorResponse>(response.body);
    } catch {
      // Not JSON, will use fallback
    }
  }

  // Extract error type
  const type = errorResponse?.__type ?? `HTTP${response.status}`;

  // Extract error message
  const message =
    errorResponse?.message ??
    (response.body || getDefaultErrorMessage(response.status));

  // Determine if error is retryable
  const retryable = isRetryableError(response.status, type);

  // Extract request ID from headers
  const requestId = extractRequestId(response.headers);

  return {
    type,
    message,
    retryable,
    requestId,
    statusCode: response.status,
  };
}

/**
 * Information about an AWS error.
 */
export interface AwsErrorInfo {
  /**
   * Error type (e.g., 'MessageRejected', 'Throttling').
   */
  type: string;

  /**
   * Human-readable error message.
   */
  message: string;

  /**
   * Whether the error is retryable.
   */
  retryable: boolean;

  /**
   * AWS request ID for debugging.
   */
  requestId?: string;

  /**
   * HTTP status code.
   */
  statusCode: number;
}

/**
 * Extract the AWS request ID from response headers.
 *
 * @param headers - Response headers
 * @returns Request ID if present
 *
 * @internal
 */
function extractRequestId(headers: Record<string, string>): string | undefined {
  // AWS uses various header names for request ID
  return (
    headers['x-amzn-requestid'] ??
    headers['x-amzn-request-id'] ??
    headers['x-amz-request-id'] ??
    undefined
  );
}

/**
 * Determine if an error is retryable based on status code and error type.
 *
 * @param statusCode - HTTP status code
 * @param errorType - AWS error type
 * @returns true if the error should be retried
 *
 * @internal
 */
function isRetryableError(statusCode: number, errorType: string): boolean {
  // 5xx errors are generally retryable (server errors)
  if (statusCode >= 500) {
    return true;
  }

  // 429 (rate limiting) is retryable
  if (statusCode === 429) {
    return true;
  }

  // Throttling errors are retryable
  if (errorType.toLowerCase().includes('throttl')) {
    return true;
  }

  // Service unavailable errors are retryable
  if (errorType.toLowerCase().includes('serviceunavailable')) {
    return true;
  }

  // Request timeout errors are retryable
  if (errorType.toLowerCase().includes('requesttimeout')) {
    return true;
  }

  // 4xx errors are generally not retryable (client errors)
  return false;
}

/**
 * Get a default error message for a status code.
 *
 * @param statusCode - HTTP status code
 * @returns Default error message
 *
 * @internal
 */
function getDefaultErrorMessage(statusCode: number): string {
  switch (statusCode) {
    case 400:
      return 'Bad Request';
    case 401:
      return 'Unauthorized';
    case 403:
      return 'Forbidden';
    case 404:
      return 'Not Found';
    case 429:
      return 'Too Many Requests';
    case 500:
      return 'Internal Server Error';
    case 502:
      return 'Bad Gateway';
    case 503:
      return 'Service Unavailable';
    case 504:
      return 'Gateway Timeout';
    default:
      return `HTTP Error ${statusCode}`;
  }
}

/**
 * Extract metadata from response headers.
 *
 * This includes information like request ID, rate limits, etc.
 *
 * @param headers - Response headers
 * @returns Response metadata
 *
 * @example
 * ```typescript
 * const headers = {
 *   'x-amzn-requestid': 'abc-123',
 *   'x-amzn-ratelimit-limit': '100',
 *   'x-amzn-ratelimit-remaining': '95'
 * };
 *
 * const metadata = extractResponseMetadata(headers);
 * console.log('Request ID:', metadata.requestId);
 * console.log('Rate limit:', metadata.rateLimit);
 * ```
 */
export function extractResponseMetadata(headers: Record<string, string>): ResponseMetadata {
  return {
    requestId: extractRequestId(headers),
    rateLimit: headers['x-amzn-ratelimit-limit']
      ? parseInt(headers['x-amzn-ratelimit-limit'], 10)
      : undefined,
    rateLimitRemaining: headers['x-amzn-ratelimit-remaining']
      ? parseInt(headers['x-amzn-ratelimit-remaining'], 10)
      : undefined,
    retryAfter: headers['retry-after']
      ? parseInt(headers['retry-after'], 10)
      : undefined,
  };
}

/**
 * Metadata extracted from response headers.
 */
export interface ResponseMetadata {
  /**
   * AWS request ID for debugging.
   */
  requestId?: string;

  /**
   * Rate limit for this endpoint.
   */
  rateLimit?: number;

  /**
   * Remaining requests before rate limit.
   */
  rateLimitRemaining?: number;

  /**
   * Seconds to wait before retrying (for 429 responses).
   */
  retryAfter?: number;
}

File: aws/ses/rust/src/http/response.rs
//! HTTP response handling for AWS SES API.
//!
//! This module provides response parsing and error extraction for AWS SES API calls.

use http::StatusCode;
use serde::de::DeserializeOwned;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

use crate::error::{SesError, SesResult};

/// A response from the AWS SES API.
///
/// This struct wraps the HTTP response and provides convenient methods
/// for parsing JSON data and extracting metadata.
#[derive(Debug, Clone)]
pub struct SesResponse {
    /// HTTP status code
    status: StatusCode,

    /// Response headers
    headers: HashMap<String, String>,

    /// Response body
    body: Vec<u8>,

    /// AWS request ID (from x-amzn-RequestId header)
    request_id: Option<String>,
}

impl SesResponse {
    /// Create a new response.
    ///
    /// # Arguments
    ///
    /// * `status` - HTTP status code
    /// * `headers` - Response headers
    /// * `body` - Response body
    ///
    /// # Examples
    ///
    /// ```rust
    /// use integrations_aws_ses::http::SesResponse;
    /// use http::StatusCode;
    /// use std::collections::HashMap;
    ///
    /// let response = SesResponse::new(
    ///     StatusCode::OK,
    ///     HashMap::new(),
    ///     b"response body".to_vec()
    /// );
    /// ```
    pub fn new(
        status: StatusCode,
        headers: HashMap<String, String>,
        body: Vec<u8>,
    ) -> Self {
        // Extract request ID from headers
        let request_id = headers
            .get("x-amzn-requestid")
            .or_else(|| headers.get("x-amzn-request-id"))
            .or_else(|| headers.get("x-amz-request-id"))
            .cloned();

        Self {
            status,
            headers,
            body,
            request_id,
        }
    }

    /// Create a response from a reqwest Response.
    ///
    /// # Arguments
    ///
    /// * `response` - The reqwest response
    ///
    /// # Returns
    ///
    /// A new `SesResponse` or an error if reading the body fails.
    ///
    /// # Examples
    ///
    /// ```rust,ignore
    /// use integrations_aws_ses::http::SesResponse;
    /// use reqwest::Response;
    ///
    /// async fn from_reqwest(response: Response) -> Result<SesResponse, Box<dyn std::error::Error>> {
    ///     let ses_response = SesResponse::from_reqwest(response).await?;
    ///     Ok(ses_response)
    /// }
    /// ```
    pub async fn from_reqwest(response: reqwest::Response) -> SesResult<Self> {
        let status = response.status();

        // Extract headers
        let mut headers = HashMap::new();
        for (name, value) in response.headers() {
            if let Ok(value_str) = value.to_str() {
                headers.insert(name.as_str().to_lowercase(), value_str.to_string());
            }
        }

        // Read body
        let body = response
            .bytes()
            .await
            .map_err(|e| SesError::Transport {
                message: format!("Failed to read response body: {}", e),
                source: Some(Box::new(e)),
                retryable: true,
            })?
            .to_vec();

        Ok(Self::new(status, headers, body))
    }

    /// Get the HTTP status code.
    ///
    /// # Examples
    ///
    /// ```rust
    /// use integrations_aws_ses::http::SesResponse;
    /// use http::StatusCode;
    /// use std::collections::HashMap;
    ///
    /// let response = SesResponse::new(StatusCode::OK, HashMap::new(), vec![]);
    /// assert_eq!(response.status(), StatusCode::OK);
    /// ```
    pub fn status(&self) -> StatusCode {
        self.status
    }

    /// Get a header value.
    ///
    /// # Arguments
    ///
    /// * `name` - The header name (case-insensitive)
    ///
    /// # Returns
    ///
    /// The header value if present, None otherwise.
    ///
    /// # Examples
    ///
    /// ```rust
    /// use integrations_aws_ses::http::SesResponse;
    /// use http::StatusCode;
    /// use std::collections::HashMap;
    ///
    /// let mut headers = HashMap::new();
    /// headers.insert("content-type".to_string(), "application/json".to_string());
    ///
    /// let response = SesResponse::new(StatusCode::OK, headers, vec![]);
    /// assert_eq!(response.header("content-type"), Some("application/json"));
    /// ```
    pub fn header(&self, name: &str) -> Option<&str> {
        self.headers.get(&name.to_lowercase()).map(|s| s.as_str())
    }

    /// Get all headers.
    ///
    /// # Examples
    ///
    /// ```rust
    /// use integrations_aws_ses::http::SesResponse;
    /// use http::StatusCode;
    /// use std::collections::HashMap;
    ///
    /// let mut headers = HashMap::new();
    /// headers.insert("content-type".to_string(), "application/json".to_string());
    ///
    /// let response = SesResponse::new(StatusCode::OK, headers.clone(), vec![]);
    /// assert_eq!(response.headers(), &headers);
    /// ```
    pub fn headers(&self) -> &HashMap<String, String> {
        &self.headers
    }

    /// Get the response body as bytes.
    ///
    /// # Examples
    ///
    /// ```rust
    /// use integrations_aws_ses::http::SesResponse;
    /// use http::StatusCode;
    /// use std::collections::HashMap;
    ///
    /// let body = b"response body".to_vec();
    /// let response = SesResponse::new(StatusCode::OK, HashMap::new(), body.clone());
    /// assert_eq!(response.body(), &body);
    /// ```
    pub fn body(&self) -> &[u8] {
        &self.body
    }

    /// Get the response body as a UTF-8 string.
    ///
    /// # Returns
    ///
    /// The body as a string, or an error if it's not valid UTF-8.
    ///
    /// # Examples
    ///
    /// ```rust
    /// use integrations_aws_ses::http::SesResponse;
    /// use http::StatusCode;
    /// use std::collections::HashMap;
    ///
    /// let response = SesResponse::new(
    ///     StatusCode::OK,
    ///     HashMap::new(),
    ///     b"response body".to_vec()
    /// );
    /// assert_eq!(response.body_string().unwrap(), "response body");
    /// ```
    pub fn body_string(&self) -> SesResult<&str> {
        std::str::from_utf8(&self.body).map_err(|e| SesError::Serialization {
            message: format!("Response body is not valid UTF-8: {}", e),
        })
    }

    /// Parse the response body as JSON.
    ///
    /// # Returns
    ///
    /// The deserialized JSON value or an error if parsing fails.
    ///
    /// # Examples
    ///
    /// ```rust
    /// use integrations_aws_ses::http::SesResponse;
    /// use http::StatusCode;
    /// use std::collections::HashMap;
    /// use serde::Deserialize;
    ///
    /// #[derive(Deserialize)]
    /// struct ApiResponse {
    ///     message: String,
    /// }
    ///
    /// # fn example() -> Result<(), Box<dyn std::error::Error>> {
    /// let body = r#"{"message": "success"}"#.as_bytes().to_vec();
    /// let response = SesResponse::new(StatusCode::OK, HashMap::new(), body);
    ///
    /// let data: ApiResponse = response.json()?;
    /// assert_eq!(data.message, "success");
    /// # Ok(())
    /// # }
    /// ```
    pub fn json<T: DeserializeOwned>(&self) -> SesResult<T> {
        serde_json::from_slice(&self.body).map_err(Into::into)
    }

    /// Get the AWS request ID.
    ///
    /// This can be used for debugging and support requests.
    ///
    /// # Examples
    ///
    /// ```rust
    /// use integrations_aws_ses::http::SesResponse;
    /// use http::StatusCode;
    /// use std::collections::HashMap;
    ///
    /// let mut headers = HashMap::new();
    /// headers.insert("x-amzn-requestid".to_string(), "abc-123".to_string());
    ///
    /// let response = SesResponse::new(StatusCode::OK, headers, vec![]);
    /// assert_eq!(response.request_id(), Some("abc-123"));
    /// ```
    pub fn request_id(&self) -> Option<&str> {
        self.request_id.as_deref()
    }

    /// Check if the response indicates success.
    ///
    /// # Returns
    ///
    /// `true` if the status code is in the 2xx range.
    ///
    /// # Examples
    ///
    /// ```rust
    /// use integrations_aws_ses::http::SesResponse;
    /// use http::StatusCode;
    /// use std::collections::HashMap;
    ///
    /// let response = SesResponse::new(StatusCode::OK, HashMap::new(), vec![]);
    /// assert!(response.is_success());
    ///
    /// let error_response = SesResponse::new(StatusCode::BAD_REQUEST, HashMap::new(), vec![]);
    /// assert!(!error_response.is_success());
    /// ```
    pub fn is_success(&self) -> bool {
        self.status.is_success()
    }

    /// Extract error information from the response.
    ///
    /// This parses AWS error responses and converts them to SesError.
    ///
    /// # Returns
    ///
    /// A `SesError` with error details from the response.
    ///
    /// # Examples
    ///
    /// ```rust
    /// use integrations_aws_ses::http::SesResponse;
    /// use http::StatusCode;
    /// use std::collections::HashMap;
    ///
    /// # fn example() -> Result<(), Box<dyn std::error::Error>> {
    /// let error_body = r#"{
    ///     "__type": "MessageRejected",
    ///     "message": "Email address is not verified"
    /// }"#.as_bytes().to_vec();
    ///
    /// let response = SesResponse::new(StatusCode::BAD_REQUEST, HashMap::new(), error_body);
    /// let error = response.into_error();
    /// # Ok(())
    /// # }
    /// ```
    pub fn into_error(self) -> SesError {
        // Try to parse as JSON error response
        if let Ok(error_response) = serde_json::from_slice::<AwsErrorResponse>(&self.body) {
            return self.error_from_aws_response(error_response);
        }

        // Fallback to generic error based on status code
        self.error_from_status_code()
    }

    /// Convert AWS error response to SesError.
    fn error_from_aws_response(&self, error: AwsErrorResponse) -> SesError {
        let code = error.type_field.unwrap_or_else(|| "Unknown".to_string());
        let message = error.message.unwrap_or_else(|| "Unknown error".to_string());

        // Determine if the error is retryable
        let retryable = match self.status.as_u16() {
            500..=599 => true, // Server errors are retryable
            429 => true,       // Rate limiting is retryable
            _ => false,
        };

        SesError::AwsApi {
            code,
            message,
            request_id: self.request_id.clone(),
            retryable,
        }
    }

    /// Convert status code to SesError.
    fn error_from_status_code(&self) -> SesError {
        let message = self.body_string().unwrap_or("Unknown error").to_string();
        let retryable = self.status.is_server_error() || self.status == StatusCode::TOO_MANY_REQUESTS;

        SesError::AwsApi {
            code: self.status.as_u16().to_string(),
            message,
            request_id: self.request_id.clone(),
            retryable,
        }
    }

    /// Extract pagination token from response.
    ///
    /// Many AWS SES API operations support pagination. This method extracts
    /// the NextToken field if present.
    ///
    /// # Returns
    ///
    /// The next token for pagination, if present.
    ///
    /// # Examples
    ///
    /// ```rust
    /// use integrations_aws_ses::http::SesResponse;
    /// use http::StatusCode;
    /// use std::collections::HashMap;
    ///
    /// # fn example() -> Result<(), Box<dyn std::error::Error>> {
    /// let body = r#"{
    ///     "Identities": [],
    ///     "NextToken": "abc123"
    /// }"#.as_bytes().to_vec();
    ///
    /// let response = SesResponse::new(StatusCode::OK, HashMap::new(), body);
    /// let token = response.pagination_token()?;
    /// assert_eq!(token, Some("abc123".to_string()));
    /// # Ok(())
    /// # }
    /// ```
    pub fn pagination_token(&self) -> SesResult<Option<String>> {
        #[derive(Deserialize)]
        struct PaginationResponse {
            #[serde(rename = "NextToken")]
            next_token: Option<String>,
        }

        if self.body.is_empty() {
            return Ok(None);
        }

        match serde_json::from_slice::<PaginationResponse>(&self.body) {
            Ok(resp) => Ok(resp.next_token),
            Err(_) => Ok(None), // No pagination token present
        }
    }
}

/// AWS error response structure.
///
/// This represents the JSON structure of AWS API error responses.
#[derive(Debug, Clone, Deserialize, Serialize)]
struct AwsErrorResponse {
    /// Error type (e.g., "MessageRejected", "Throttling")
    #[serde(rename = "__type")]
    type_field: Option<String>,

    /// Error message
    #[serde(rename = "message")]
    message: Option<String>,

    /// Additional error details
    #[serde(flatten)]
    extra: HashMap<String, serde_json::Value>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_ses_response_new() {
        let response = SesResponse::new(
            StatusCode::OK,
            HashMap::new(),
            b"test body".to_vec(),
        );

        assert_eq!(response.status(), StatusCode::OK);
        assert_eq!(response.body(), b"test body");
    }

    #[test]
    fn test_ses_response_header() {
        let mut headers = HashMap::new();
        headers.insert("content-type".to_string(), "application/json".to_string());

        let response = SesResponse::new(StatusCode::OK, headers, vec![]);

        assert_eq!(response.header("content-type"), Some("application/json"));
        assert_eq!(response.header("Content-Type"), Some("application/json"));
        assert_eq!(response.header("missing"), None);
    }

    #[test]
    fn test_ses_response_body_string() {
        let response = SesResponse::new(
            StatusCode::OK,
            HashMap::new(),
            b"test body".to_vec(),
        );

        assert_eq!(response.body_string().unwrap(), "test body");
    }

    #[test]
    fn test_ses_response_json() {
        #[derive(Deserialize)]
        struct TestData {
            name: String,
            value: i32,
        }

        let body = r#"{"name": "test", "value": 42}"#.as_bytes().to_vec();
        let response = SesResponse::new(StatusCode::OK, HashMap::new(), body);

        let data: TestData = response.json().unwrap();
        assert_eq!(data.name, "test");
        assert_eq!(data.value, 42);
    }

    #[test]
    fn test_ses_response_request_id() {
        let mut headers = HashMap::new();
        headers.insert("x-amzn-requestid".to_string(), "abc-123".to_string());

        let response = SesResponse::new(StatusCode::OK, headers, vec![]);

        assert_eq!(response.request_id(), Some("abc-123"));
    }

    #[test]
    fn test_ses_response_request_id_alternate_headers() {
        let mut headers = HashMap::new();
        headers.insert("x-amz-request-id".to_string(), "xyz-789".to_string());

        let response = SesResponse::new(StatusCode::OK, headers, vec![]);

        assert_eq!(response.request_id(), Some("xyz-789"));
    }

    #[test]
    fn test_ses_response_is_success() {
        let success = SesResponse::new(StatusCode::OK, HashMap::new(), vec![]);
        assert!(success.is_success());

        let error = SesResponse::new(StatusCode::BAD_REQUEST, HashMap::new(), vec![]);
        assert!(!error.is_success());
    }

    #[test]
    fn test_ses_response_into_error_with_json() {
        let error_body = r#"{
            "__type": "MessageRejected",
            "message": "Email address is not verified"
        }"#
        .as_bytes()
        .to_vec();

        let response = SesResponse::new(StatusCode::BAD_REQUEST, HashMap::new(), error_body);
        let error = response.into_error();

        assert_eq!(error.error_code(), Some("MessageRejected"));
        assert!(!error.is_retryable());
    }

    #[test]
    fn test_ses_response_into_error_without_json() {
        let response = SesResponse::new(
            StatusCode::INTERNAL_SERVER_ERROR,
            HashMap::new(),
            b"Internal error".to_vec(),
        );
        let error = response.into_error();

        assert!(error.is_retryable());
    }

    #[test]
    fn test_ses_response_pagination_token() {
        let body = r#"{
            "Identities": [],
            "NextToken": "abc123"
        }"#
        .as_bytes()
        .to_vec();

        let response = SesResponse::new(StatusCode::OK, HashMap::new(), body);
        let token = response.pagination_token().unwrap();

        assert_eq!(token, Some("abc123".to_string()));
    }

    #[test]
    fn test_ses_response_pagination_token_missing() {
        let body = r#"{
            "Identities": []
        }"#
        .as_bytes()
        .to_vec();

        let response = SesResponse::new(StatusCode::OK, HashMap::new(), body);
        let token = response.pagination_token().unwrap();

        assert_eq!(token, None);
    }

    #[test]
    fn test_ses_response_pagination_token_empty_body() {
        let response = SesResponse::new(StatusCode::OK, HashMap::new(), vec![]);
        let token = response.pagination_token().unwrap();

        assert_eq!(token, None);
    }
}
